versioned_docs/version-2.0/tutorials/rag.mdx

```diff
@@ -0,0 +1,545 @@
+    input_question = example.inputs["input_question"]
+    prediction = run.outputs["answer"]
+
+    # LLM grader
+    llm = ChatOpenAI(model="gpt-4-turbo", temperature=0)
+
+    # Structured prompt
+    answer_grader = grade_prompt_answer_helpfulness | llm
+
+    # Run evaluator
+    score = answer_grader.invoke({"question": input_question,
+                                  "student_answer": prediction})
+    score = score["Score"]
+
+    return {"key": "answer_helpfulness_score", "score": score}
+```
+
+```python
+experiment_results = evaluate(
+    predict_rag_answer,
+    data=dataset_name,
+    evaluators=[answer_helpfulness_evaluator],
+    experiment_prefix="rag-answer-helpfulness",
+    metadata={"version": "LCEL context, gpt-4-0125-preview"},
+)
+```
+
+    View the evaluation results for experiment: 'rag-answer-helpfulness-e64d4893' at:
+    https://smith.langchain.com/o/ebbaf2eb-769b-4505-aca2-d11de10372a4/datasets/5f34de91-78ab-4ae2-83f3-8f7f93afc998/compare?selectedSessions=60ee674a-fb59-4c68-86fe-148355fc659e
+
+
+
+
+
+    0it [00:00, ?it/s]
+
+### **Response vs retrieved docs**
+
+Here is an example prompt that we can use:
+
+https://smith.langchain.com/hub/langchain-ai/rag-answer-hallucination
+
+Here is the information flow:
+
+![answer_hallucination_flow.png](./static/answer_hallucination_flow.png)
+
+```python
+# Prompt
+grade_prompt_hallucinations = prompt = hub.pull("langchain-ai/rag-answer-hallucination")
+
+def answer_hallucination_evaluator(run, example) -> dict:
+    """
+    A simple evaluator for generation hallucination
+    """
+
+    # RAG inputs
+    input_question = example.inputs["input_question"]
+    contexts = run.outputs["contexts"]
+
+    # RAG answer
+    prediction = run.outputs["answer"]
+
+    # LLM grader
+    llm = ChatOpenAI(model="gpt-4-turbo", temperature=0)
+
+    # Structured prompt
+    answer_grader = grade_prompt_hallucinations | llm
+
+    # Get score
+    score = answer_grader.invoke({"documents": contexts,
+                                  "student_answer": prediction})
+    score = score["Score"]
+
+    return {"key": "answer_hallucination", "score": score}
+```
+
+```python
+experiment_results = evaluate(
+    predict_rag_answer_with_context,
+    data=dataset_name,
+    evaluators=[answer_hallucination_evaluator],
+    experiment_prefix="rag-answer-hallucination",
+    metadata={"version": "LCEL context, gpt-4-0125-preview"},
+)
+```
+
+    View the evaluation results for experiment: 'rag-answer-hallucination-0bcf36cb' at:
+    https://smith.langchain.com/o/ebbaf2eb-769b-4505-aca2-d11de10372a4/datasets/5f34de91-78ab-4ae2-83f3-8f7f93afc998/compare?selectedSessions=db391b03-d4b2-4c57-9e1c-8ef33389c60d
+
+
+
+
+
+    0it [00:00, ?it/s]
+
+### **Retrieved docs vs input**
+
+Here is an example prompt that we can use:
+
+https://smith.langchain.com/hub/langchain-ai/rag-document-relevance
+
+Here is the information flow:
+
+![document_relevance_flow.png](./static/document_relevance_flow.png)
```

Comment:
To me these diagrams (specifically the information flow ones) are confusing and I can't really grok out what they mean